CAToolsetPerfDataCollector/PerfDataCollector.h

```diff
@@ -0,0 +1,132 @@
+    microseconds EspXAllChecks;             // Analysis of functions by EspXEngine and its extensions
+    microseconds EspXPathSensitiveChecks;   // Path-sensitive analysis of functions by EspXEngine and its extensions
+
+    // This header is used to print the performance data to the console.
+    // It should be updated if the PerfData struct is updated.
+    // Note that this header does not include the file path, which should be printed separately if needed.
+    static constexpr std::wstring_view PerfDataHeader =
+        L"Front End Pass, BackEnd Pass, Code Analysis Pass, "
+        L"AST Creation, AST Clients, "                                      // Children of CodeAnalysis Pass
+        L"Function Analysis, "                                              // Child of AST Clients
+        L"FPA Function Analysis, EspX CFG Build, EspX Function Analysis, "  // Children of Function Analysis
+        L"EspX Path-sensitive Analysis";                                    // Child of EspX Function Analysis
+};
+
+class PerfDataCollector : public IAnalyzer
+{
+private:
+    std::unordered_map<std::wstring, PathId> filePaths{};   // List of file paths
+    std::map<PathId, PerfData> perfDataPerTu{};                  // Performance data per TU
+
+    // Find the PathId for the given file path.
+    // If the file path is not found, a new PathId is created.
+    PathId& FindFilePathId(std::wstring path)
+    {
+        static int id = 0;
+        auto result = filePaths.try_emplace(path, PathId{});
+        if (result.second)
+        {
+            result.first->second.id = id++;
+            result.first->second.path = &result.first->first;
+        }
+        return result.first->second;
+    }
+
+    // Gets duration of the specified compiler pass (should be one of FE, BE, and CA pass),
+    // finds or adds performance data for the file path, and adds the duration to the specified field
+    // of the performance data.
+    template<typename TPass, microseconds PerfData::* RecField>
+    void GetPassDuration(TPass pass)
+    {
+        static_assert(std::is_base_of_v<CompilerPass, TPass>, "TPass must be derived from CompilerPass");
+
+        const auto tuPathId = FindFilePathId(pass.InputSourcePath());
+
+        if (verbose_)
+        {
+            std::wcout << L"[" << *tuPathId.path << L"](" << pass.EventName() << L") = "
+                << std::chrono::duration_cast<microseconds>(pass.Duration()).count() << L"(microsec)" << std::endl;
+        }
+
+        auto& data = perfDataPerTu[tuPathId];
+        data.*RecField += std::chrono::duration_cast<microseconds>(pass.Duration());
+    }
+
+    // Gets duration of the specified CA activity (should be a descendent of CA Pass),
+    // finds performance data for the file path, and adds the duration to the specified field
+    // of the performance data.
+    template<typename TEvent, microseconds PerfData::* RecField>
+    void GetCodeAnalysisEventDuration(CodeAnalysisPass pass, TEvent event)
```

Comment:
The implementation looks really similar to `GetPassDuration`. Is it possible to only have one of them (possibly something like GetActivityDuration)?

Reply:
This one requires CAPass and Event. I would like to leave them separate.